src/components/ble/AppleNotificationCenterClient.cpp

```diff
@@ -0,0 +1,514 @@
+      if (controlPointDescriptorHandle == 0) {
+        NRF_LOG_INFO("ANCS Descriptor discovered : %d", descriptor->handle);
+        controlPointDescriptorHandle = descriptor->handle;
+        isControlDescriptorFound = true;
+      }
+    } else if (characteristicValueHandle == dataSourceHandle && ble_uuid_cmp(&dataSourceChar.u, &descriptor->uuid.u)) {
+      if (dataSourceDescriptorHandle == 0) {
+        NRF_LOG_INFO("ANCS Descriptor discovered : %d", descriptor->handle);
+        dataSourceDescriptorHandle = descriptor->handle;
+        isDataDescriptorFound = true;
+        uint8_t value[2] {1, 0};
+        ble_gattc_write_flat(connectionHandle, dataSourceDescriptorHandle, value, sizeof(value), NewAlertSubcribeCallback, this);
+      }
+    }
+  } else {
+    if (error->status != BLE_HS_EDONE) {
+      char errorStr[55];
+      snprintf(errorStr, sizeof(errorStr), "ANCS Descriptor discovery ERROR: %d", error->status);
+      NRF_LOG_INFO(errorStr);
+    }
+    if (isDescriptorFound == isDataDescriptorFound)
+      onServiceDiscovered(connectionHandle);
+  }
+  return 0;
+}
+
+int AppleNotificationCenterClient::OnNewAlertSubcribe(uint16_t connectionHandle,
+                                                      const ble_gatt_error* error,
+                                                      ble_gatt_attr* /*attribute*/) {
+  if (error->status == 0) {
+    NRF_LOG_INFO("ANCS New alert subscribe OK");
+  } else {
+    NRF_LOG_INFO("ANCS New alert subscribe ERROR");
+  }
+  if (isDescriptorFound == isControlDescriptorFound && isDescriptorFound == isDataDescriptorFound)
+    onServiceDiscovered(connectionHandle);
+
+  return 0;
+}
+
+int AppleNotificationCenterClient::OnControlPointWrite(uint16_t /*connectionHandle*/,
+                                                       const ble_gatt_error* error,
+                                                       ble_gatt_attr* /*attribute*/) {
+  if (error->status == 0) {
+    NRF_LOG_INFO("ANCS Control Point write OK");
+  } else {
+    char errorStr[55];
+    snprintf(errorStr, sizeof(errorStr), "ANCS Control Point ERROR: %d", error->status);
+    NRF_LOG_INFO(errorStr);
+  }
+  return 0;
+}
+
+void AppleNotificationCenterClient::OnNotification(ble_gap_event* event) {
+  if (event->notify_rx.attr_handle == notificationSourceHandle || event->notify_rx.attr_handle == notificationSourceDescriptorHandle) {
+    NRF_LOG_INFO("ANCS Notification received");
+    uint8_t eventId;
+    uint8_t eventFlags;
+    uint8_t category;
+    uint8_t categoryCount;
+    uint32_t notificationUuid;
+
+    os_mbuf_copydata(event->notify_rx.om, 0, 1, &eventId);
+    os_mbuf_copydata(event->notify_rx.om, 1, 1, &eventFlags);
+    os_mbuf_copydata(event->notify_rx.om, 2, 1, &category);
+    os_mbuf_copydata(event->notify_rx.om, 3, 1, &categoryCount);
+    os_mbuf_copydata(event->notify_rx.om, 4, 4, &notificationUuid);
```

Comment:
Maybe gcc is smart enough to get this, but I will aks anyway :smile:
You create those local variables only to insert them to a notification. Couldn't you cut the local vairables and `os_mbuf_copydata` directly into the `ancsNotif`?